Sources/core/objects/Projectiles.cpp

```diff
@@ -0,0 +1,233 @@
+#include "Projectiles.h"
+#include "../Helpers.h"
+
+std::ostream& operator<<(std::stringstream& file,const std::vector<Dir>& s)
+{
+    file << " " << s.size() << " ";
+    for (auto it : s)
+        file << " " << it << " ";
+    return file;
+}
+
+std::istream& operator>>(std::stringstream& file,std::vector<Dir>& s)
+{
+    size_t size;
+    file >> size;
+    s.reserve(size);
+
+    unsigned int local_id;
+    for (size_t i = 0; i < size; ++i)
+    {
+        file >> local_id;
+        s.push_back(local_id);
+    }
+    return file;
+}
+
+unsigned int hash(const std::vector<Dir>& content)
+{
+    unsigned int retval = 0;
+    int i = 1;
+    for (auto it = content.begin(); it != content.end(); ++it, ++i)
+    {
+        retval += *it * i;
+    }
+    return retval;
+}
+
+
+Projectile::Projectile(size_t id) : IMovable(id)
+{
+    SetPassable(D_ALL, Passable::EMPTY);
+    damage = 0;
+    SetSprite("icons/projectiles.dmi");
+    v_level = 5;
+    current_step_ = 0;
+    set_target_ = 0;
+    tickSpeed = 1;
+    pixSpeed = 4;
+    reached_target = 0;
+}
+void Projectile::Process()
+{
+   if (set_target_ == 1)
+   {
+       Dir step =  movement_[current_step_];
+       Rotate(step);
+       if(!CheckPassable())
+       {
+           return;
+       }
+       else if (reached_target)
+       {
+           MainMove();
+           return;
+       }
+       else if (current_step_ < movement_.size()-1 && !reached_target)
+       {
+           current_step_++;
+       }
+       else if(!reached_target && current_step_ == movement_.size()-1)
+       {
+           current_step_ =0;
+           Dir tmp = movement_[1];
+           movement_.clear();
+           movement_.push_back(tmp);
+           reached_target = 1;
+       }    
+       MainMove();
+   }
+}
+
+bool Projectile::CheckPassable()
+{
+    if (IMovable::CheckPassable())
+    {
+        return true;
+    }
+    this->Delete();
+    return false;
+}
+//TODO maybe it can apply froce depending on bullet speed but that should be considered        ApplyForce(DirToVDir[m->dMove]);
+void Projectile::MakeMovementPattern(VDir target,id_ptr_on<Human> someone)
+{
+    set_target_=1;
+    int x = target.x;
+    int y = target.y;
+    int abs_x = std::abs(x);
+    int abs_y = std::abs(y);
+    shooter_ = someone;
+    Dir facing = someone->GetDir();
+    if (abs_x == abs_y)
+    {
+        if (y<0)
+        {
+            if (facing == 0|| facing == 1)
+            {
```

Comment:
I am not sure that you mean, but I think you can just create projectile not on the same tile with shooter, but on the next to him and that should work well.